uploader/uploader.go

```diff
@@ -1,149 +1,137 @@
-		copyUploaders = append(copyUploaders, NewNoopUploader())
+	if len(config.S3Uploaders) > 0 {
+		bucketName = config.S3Uploaders[0]
 	}
 
-	return copyUploaders
-}
+	if param, ok := titusInfo.GetPassthroughAttributes()[s3BucketNameParam]; ok {
+		bucketName = param
+		useTitusRole = false
+	}
 
-// Uploaders is a slice wrapper that contains all of the
-// uploaders to use when tasks complete
-type Uploaders struct {
-	uploaders []Uploader
-}
+	if param, ok := titusInfo.GetPassthroughAttributes()[s3PathPrefixParam]; ok {
+		pathPrefix = param
+	}
 
-// NewUploaders creates a new instance of an Uploaders object
-func NewUploaders(config *config.Config, m metrics.Reporter) (*Uploaders, error) {
-	e := &Uploaders{}
+	if bucketName != "" {
+		s3, err := NewS3Backend(m, bucketName, pathPrefix, titusInfo.GetIamProfile(), taskID, useTitusRole)
+		if err != nil {
+			return nil, err
+		}
+		return NewUploaderWithBackend(s3), nil
+	}
 
-	e.uploaders = append(e.uploaders, collectCopyUploaders(config)...)
-	e.uploaders = append(e.uploaders, collectS3Uploaders(config, m)...)
-	e.uploaders = append(e.uploaders, collectNoopUploaders(config)...)
+	if len(config.CopyUploaders) > 0 {
+		return NewUploaderWithBackend(NewCopyBackend(config.CopyUploaders[0])), nil
+	}
 
-	return e, nil
+	return NewUploaderWithBackend(NewNoopBackend()), nil
 }
 
-// NewUploadersFromUploaderArray creates a new instance of an Uploaders object from a list of Uploader instances
-func NewUploadersFromUploaderArray(uploaders []Uploader) *Uploaders {
-	e := &Uploaders{
-		uploaders: make([]Uploader, len(uploaders)),
-	}
-	copy(e.uploaders, uploaders)
-	return e
+func NewUploaderWithBackend(backend Backend) *Uploader {
+	return &Uploader{backend}
 }
 
-// Performs a parallel upload of all of files in a directory but
-// not its subdirectories. A slice containing the error results for
-// each upload with an error is returned.
-func uploadDir(ctx context.Context, uploader Uploader, local string, remote string, ctypeFunc ContentTypeInferenceFunction) []error {
-	var errs []error
+// Upload of all of files in a directory but not its subdirectories. Performs the uploads in parallel.
+func uploadDir(ctx context.Context, uploader Backend, local string, remote string, ctypeFunc ContentTypeInferenceFunction) error {
+	var errs *multierror.Error
 
 	fi, err := os.Stat(local)
 	if err != nil {
-		return append(errs, err)
+		errs = multierror.Append(errs, err)
 	}
 
 	if fi.IsDir() {
 		finfos, err := ioutil.ReadDir(local)
 		if err != nil {
-			return append(errs, err)
-		}
-		// Iterate over each file and upload it
-		for _, finfo := range finfos {
-			if finfo.IsDir() {
-				continue // don't upload subdirs
+			errs = multierror.Append(errs, err)
+		} else {
+			uploadErrs := make([]error, len(finfos))
+
+			// Iterate over each file and upload it
+			var wg sync.WaitGroup
+
+			for i, finfo := range finfos {
+				if finfo.IsDir() {
+					continue // don't upload subdirs
+				}
+
+				wg.Add(1)
```

Comment:
I don't know if it's legal for multiple goroutines to simultaneously write to the same slice. I think you have to use a channel to collect the errors.